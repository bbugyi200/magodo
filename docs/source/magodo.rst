magodo package
==============

.. automodule:: magodo
   :members:
   :undoc-members:
   :show-inheritance:

Submodules
----------

.. toctree::
   :maxdepth: 4

   magodo.spells
   magodo.types
