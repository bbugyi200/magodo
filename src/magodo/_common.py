"""Shared utility functions / constants / classes."""

from __future__ import annotations

from typing import Final

from .types import Priority


DEFAULT_PRIORITY: Final[Priority] = "O"
PUNCTUATION: Final = ",.?!;"
