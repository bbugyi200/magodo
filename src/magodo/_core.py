"""The magodo package's catch-all module.

You should only add code to this module when you are unable to find ANY other
module to add it to.
"""

from __future__ import annotations


def dummy(x: int, y: int) -> int:
    """Dummy function."""
    return x + y
