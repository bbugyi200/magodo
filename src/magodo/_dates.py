"""Datetime utilities used throughout this library."""

from __future__ import annotations

import datetime as dt
from typing import Final


DATE_FMT: Final = "%Y-%m-%d"
RE_DATE: Final = r"[1-9][0-9][0-9][0-9]-[0-1][0-9]-[0-3][0-9]"


def to_date(yyyymmdd: str) -> dt.date:
    """Helper function for constructing a date object."""
    return dt.datetime.strptime(yyyymmdd, DATE_FMT).date()
