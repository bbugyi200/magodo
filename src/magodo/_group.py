"""Contains the TodoGroup class definition."""

from __future__ import annotations

from dataclasses import dataclass
import datetime as dt
from logging import Logger
from pathlib import Path
from typing import Generic, Iterable, Iterator, Type

from eris import Err
from metaman import cname
from typist import PathLike

from .types import MetadataFunc, Priority, T


logger = Logger(__name__)


@dataclass
class MetadataCheck:
    """Wrapper for the MetadataFunc type."""

    key: str
    check: MetadataFunc = lambda _: True
    required: bool = True


class TodoGroup(Generic[T]):
    """Manages a group of Todo objects."""

    def __init__(
        self,
        todos: Iterable[T],
        todo_map: dict[str, T],
        path_map: dict[str, Path],
    ) -> None:
        self.todos = list(todos)
        self.todo_map = todo_map
        self.path_map = path_map

    def __repr__(self) -> str:  # noqa: D105
        return f"{cname(self)}({self.todos})"

    def __iter__(self) -> Iterator[T]:
        """Yields the Todo objects that belong to this group."""
        yield from self.todos

    def __len__(self) -> int:  # noqa: D105
        return len(self.todos)

    @classmethod
    def from_path(
        cls,
        todo_type: Type[T],
        path_like: PathLike,
        *,
        path_map: dict[str, Path] = None,
        todo_map: dict[str, T] = None,
    ) -> TodoGroup:
        """Reads all todo lines from a given file or directory (recursively).

        Pre-conditions:
            * `path_like` exists and is either a file or directory.
        """
        path = Path(path_like)
        if path_map is None:
            path_map = {}

        if todo_map is None:
            todo_map = {}

        assert path.exists(), f"The provided path does not exist: {path}"

        todos = []
        if path.is_file():
            logger.debug(
                "Attempting to load todos from text file: file=%r", str(path)
            )

            for line in path.read_text().split("\n"):
                line = line.strip()
                todo_result = todo_type.from_line(line)
                if isinstance(todo_result, Err):
                    continue

                todo = todo_result.ok()
                logger.debug("New todo loaded: todo=%r", todo)
                todos.append(todo)

                key = todo.ident
                path_map[key] = path
                todo_map[key] = todo

        else:
            assert path.is_dir(), (
                "The provided path exists but is neither a file nor a"
                f" directory: {path}"
            )

            logger.debug("Loading from directory: dir=%r", str(path))
            for other_path in path.glob("*"):
                if other_path.is_file() and other_path.suffix != ".txt":
                    continue

                todos.extend(
                    TodoGroup.from_path(
                        todo_type,
                        other_path,
                        path_map=path_map,
                        todo_map=todo_map,
                    )
                )

        return cls(todos, todo_map, path_map)

    def filter_by(
        self,
        *,
        contexts: Iterable[str] = (),
        create_date: dt.date = None,
        desc: str = None,
        done_date: dt.date = None,
        done: bool = None,
        metadata_checks: Iterable[MetadataCheck] = (),
        priorities: Iterable[Priority] = (),
        projects: Iterable[str] = (),
    ) -> TodoGroup:
        """Filter this group using one or more Todo properties."""
        todos = []
        path_map = {}
        todo_map = {}

        for key, todo in self.todo_map.items():
            skip_this_todo = False
            for ctx in contexts:
                if ctx.startswith("-") and ctx[1:] in todo.contexts:
                    skip_this_todo = True
                    break

                if ctx not in todo.contexts:
                    skip_this_todo = True
                    break

            for proj in projects:
                if proj.startswith("-") and proj[1:] in todo.projects:
                    skip_this_todo = True
                    break

                if proj not in todo.projects:
                    skip_this_todo = True
                    break

            if priorities and todo.priority not in priorities:
                continue

            if create_date is not None and todo.create_date != create_date:
                continue

            if desc is not None and desc.lower() not in todo.desc.lower():
                continue

            if done_date is not None and todo.done_date != done_date:
                continue

            if done is not None and todo.done != done:
                continue

            for mcheck in metadata_checks:
                key_not_found = mcheck.key not in todo.metadata
                if mcheck.required and key_not_found:
                    skip_this_todo = True
                    break

                if key_not_found:
                    continue

                mvalue = todo.metadata[mcheck.key]
                assert isinstance(mvalue, str)
                if not mcheck.check(mvalue):
                    skip_this_todo = True
                    break

            if skip_this_todo:
                continue

            todos.append(todo)
            path_map[key] = self.path_map[key]
            todo_map[key] = todo

        return TodoGroup(todos, todo_map, path_map)
