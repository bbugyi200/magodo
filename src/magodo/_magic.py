"""Contains the MagicTodoMixin class definition."""

from __future__ import annotations

import abc
import datetime as dt
from functools import total_ordering
import itertools as it
from typing import Any, List, Tuple, Type, TypeVar
import uuid

from eris import ErisError, Err, Ok, Result

from ._todo import Todo, TodoMixin
from .types import LineSpell, Metadata, Priority, TodoSpell, ValidateSpell


M = TypeVar("M", bound="MagicTodoMixin")


@total_ordering
class MagicTodoMixin(TodoMixin, abc.ABC):
    """Mixin class that implements the Todo protocol."""

    pre_todo_spells: List[TodoSpell] = []
    todo_spells: List[TodoSpell] = []
    post_todo_spells: List[TodoSpell] = []

    to_line_spells: List[LineSpell] = []
    from_line_spells: List[LineSpell] = []

    validate_spells: List[ValidateSpell] = []

    def __init__(self: M, todo: Todo):
        self._todo = todo
        self.todo = self.cast_todo_spells(todo)

    @property
    def ident(self) -> str:
        """Unique identifier."""
        return str(uuid.uuid4())

    @classmethod
    def from_line(cls: Type[M], line: str) -> Result[M, ErisError]:
        """Converts a string into a MagicTodo object."""
        err: Err[Any, ErisError]
        if error := cls.cast_validate_spells(line).err():
            err = Err(f"Failed spell validation for this todo: line={line!r}")
            return err.chain(error)

        line = cls.cast_from_line_spells(line)
        todo_result = Todo.from_line(line)

        if isinstance(todo_result, Err):
            err = Err(
                "Failed to construct basic Todo object inside of MagicTodo."
            )
            return err.chain(todo_result)

        todo = todo_result.ok()

        return Ok(cls(todo))

    def to_line(self: M) -> str:
        """Converts this MagicTodo back to a string."""
        line = self.todo.to_line()
        line = self.cast_to_line_spells(line)
        return line

    @classmethod
    def cast_todo_spells(cls: Type[M], todo: Todo) -> Todo:
        """Casts all spells associated with this MagicTodo on `todo`."""
        new_todo = todo.new()
        for todo_spell in it.chain(
            cls.pre_todo_spells, cls.todo_spells, cls.post_todo_spells
        ):
            new_todo = todo_spell(new_todo)

        return new_todo

    @classmethod
    def cast_from_line_spells(cls: Type[M], line: str) -> str:
        """Casts all from_line spells on `line`."""
        for line_spell in cls.from_line_spells:
            line = line_spell(line)
        return line

    def cast_to_line_spells(self: M, line: str) -> str:
        """Casts all to_line spells on `line`."""
        for line_spell in self.to_line_spells:
            line = line_spell(line)
        return line

    @classmethod
    def cast_validate_spells(
        cls: Type[M], line: str
    ) -> Result[None, ErisError]:
        """Casts all spells associated with this MagicTodo on `todo`."""
        for validate_spell in cls.validate_spells:
            new_todo_result = validate_spell(line)
            if isinstance(new_todo_result, Err):
                err: Err[Any, ErisError] = Err(
                    f"The {validate_spell.__name__!r} validation spell failed"
                    " while attempting to validate this line."
                )
                return err.chain(new_todo_result)

        return Ok(None)

    def new(self: M, **kwargs: Any) -> M:
        """Creates a new Todo using the current Todo's attrs as defaults."""
        return type(self)(self.todo.new(**kwargs))

    @property
    def contexts(self: M) -> Tuple[str, ...]:  # noqa: D102
        return self.todo.contexts

    @property
    def create_date(self: M) -> dt.date | None:  # noqa: D102
        return self.todo.create_date

    @property
    def desc(self: M) -> str:  # noqa: D102
        return self.todo.desc

    @property
    def done_date(self: M) -> dt.date | None:  # noqa: D102
        return self.todo.done_date

    @property
    def done(self: M) -> bool:  # noqa: D102
        return self.todo.done

    @property
    def metadata(self: M) -> Metadata:  # noqa: D102
        return self.todo.metadata

    @property
    def priority(self: M) -> Priority:  # noqa: D102
        return self.todo.priority

    @property
    def projects(self: M) -> Tuple[str, ...]:  # noqa: D102
        return self.todo.projects
