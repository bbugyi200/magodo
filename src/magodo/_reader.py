"""Contains the TodoReader class definition."""

from __future__ import annotations

import datetime as dt
from logging import Logger
from pathlib import Path
from typing import Generic, Iterable, Iterator, List, Type

from eris import Err
from metaman import cname
from typist import PathLike

from .types import Metadata, Priority, T


logger = Logger(__name__)


class TodoReader(Generic[T]):
    """Manages a group of Todo objects."""

    def __init__(self, todos: Iterable[T]) -> None:
        self._todos = sorted(todos)

    def __repr__(self) -> str:  # noqa: D105
        return f"{cname(self)}({self._todos})"

    def __iter__(self) -> Iterator[T]:
        """Yields the Todo objects that belong to this group."""
        yield from self._todos

    def __len__(self) -> int:  # noqa: D105
        return len(self._todos)

    @classmethod
    def from_path(cls, todo_type: Type[T], path_like: PathLike) -> TodoReader:
        """Reads all todo lines from a given file or directory (recursively).

        Pre-conditions:
            * `path_like` exists and is either a file or directory.
        """
        path = Path(path_like)

        assert path.exists(), f"The provided path does not exist: {path}"

        todos: List[T] = []
        if path.is_file():
            logger.debug(
                "Attempting to load todos from text file: file=%r", str(path)
            )

            for line in path.read_text().split("\n"):
                line = line.strip()
                todo_result = todo_type.from_line(line)
                if isinstance(todo_result, Err):
                    continue

                todo = todo_result.ok()
                todos.append(todo)
                logger.debug("New todo loaded: todo=%r", todo)
        else:
            assert path.is_dir(), (
                "The provided path exists but is neither a file nor a"
                f" directory: {path}"
            )

            logger.debug("Loading from directory: dir=%r", str(path))
            for other_path in path.glob("*"):
                if other_path.is_file() and other_path.suffix != ".txt":
                    continue

                other_todo_group = TodoReader.from_path(todo_type, other_path)
                todos.extend(other_todo_group)

        return cls(todos)

    def filter_by(
        self,
        *,
        contexts: Iterable[str] = None,
        create_date: dt.date = None,
        desc: str = None,
        done_date: dt.date = None,
        done: bool = None,
        metadata: Metadata = None,
        priority: Priority = None,
        projects: Iterable[str] = None,
    ) -> TodoReader:
        """Filter this group using one or more Todo properties."""
        todos = []

        for todo in self._todos:
            if contexts is not None and not all(
                ctx in todo.contexts for ctx in contexts
            ):
                continue

            if create_date is not None and todo.create_date != create_date:
                continue

            if desc is not None and desc.lower() not in todo.desc.lower():
                continue

            if done_date is not None and todo.done_date != done_date:
                continue

            if done is not None and todo.done != done:
                continue

            if metadata is not None and not all(
                v == todo.metadata.get(k, None) for (k, v) in metadata.items()
            ):
                continue

            if priority is not None and todo.priority != priority:
                continue

            if projects is not None and not all(
                proj in todo.projects for proj in projects
            ):
                continue

            todos.append(todo)

        return TodoReader(todos)
