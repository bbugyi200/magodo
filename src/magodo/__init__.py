"""A Python library for working with the todo.txt format."""

import logging as _logging

from . import spells, types
from ._common import DEFAULT_PRIORITY, from_date, to_date
from ._group import TodoGroup
from ._magic import MagicTodoMixin
from ._todo import Todo


__all__ = [
    "DEFAULT_PRIORITY",
    "MagicTodoMixin",
    "Todo",
    "TodoGroup",
    "from_date",
    "spells",
    "to_date",
    "types",
]

__author__ = "Bryan M Bugyi"
__email__ = "bryanbugyi34@gmail.com"
__version__ = "0.8.2"

_logging.getLogger(__name__).addHandler(_logging.NullHandler())
