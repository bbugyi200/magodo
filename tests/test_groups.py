"""Test that TodoGroup objects work as they should."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from pytest import mark, param
from syrupy.assertion import SnapshotAssertion as Snapshot

from magodo import TodoGroup

from .data import get_all_todo_paths
from .shared import MOCK_TODO_KWARGS, MagicTodo


params = mark.parametrize


@params("todo_file", get_all_todo_paths())
@params(
    "filter_kwargs",
    [
        param({}, id="no-filter-kwargs"),
        param({"desc": "Double colons"}, id="filter-kwarg-desc"),
        param({"contexts": ["high"]}, id="ctx"),
    ],
)
def test_group_from_path(
    snapshot: Snapshot, todo_file: Path, filter_kwargs: dict[str, Any]
) -> None:
    """Test the TodoGroup.from_path() function."""
    todo_group = TodoGroup.from_path(MagicTodo, todo_file)
    todo_group = todo_group.filter_by(**filter_kwargs)
    assert [
        repr(todo.new(**MOCK_TODO_KWARGS)) for todo in todo_group
    ] == snapshot
