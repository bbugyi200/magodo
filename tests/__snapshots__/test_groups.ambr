# name: test_group_from_path[ctx-bar]
  <class 'list'> [
    "MagicTodo(desc='High-priority item | @high ctime:HHMM dtime:HHMM', contexts=('high',), create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='A')",
  ]
---
# name: test_group_from_path[ctx-foo]
  <class 'list'> [
    "MagicTodo(desc='High-priority item | @high ctime:HHMM dtime:HHMM', contexts=('high',), create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='A')",
    "MagicTodo(desc='High-priority item. | @high ctime:HHMM dtime:HHMM', contexts=('high',), create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='A')",
  ]
---
# name: test_group_from_path[filter-kwarg-desc-bar]
  <class 'list'> [
  ]
---
# name: test_group_from_path[filter-kwarg-desc-foo]
  <class 'list'> [
    "MagicTodo(desc='Double colons should be::ignored. | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
  ]
---
# name: test_group_from_path[no-filter-kwargs-bar]
  <class 'list'> [
    "MagicTodo(desc='This todo came from the C.txt file. | +C.txt ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, projects=('C.txt',))",
    "MagicTodo(desc='Some other done todo. | @done ctime:HHMM dtime:HHMM', contexts=('done',), create_date=1900-01-01, done_date=2022-02-07, done=True, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='And another... | +another ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='B', projects=('another',))",
    "MagicTodo(desc='Then a few lines of nonesense...... | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='Then a few lines of nonesense...... | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='Then a few lines of nonesense...... | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='Then a few lines of nonesense...... | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='Then a few lines of nonesense...... | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='Then a few lines of nonesense...... | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='Then a few lines of nonesense...... | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='Then a few lines of nonesense...... | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='Then a few lines of nonesense...... | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='Then a few lines of nonesense...... | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='And +finally... one:more!', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='C', projects=('finally',))",
    "MagicTodo(desc='I do lots of stuff in this file. Some lines are todo tasks, but some lines are headers. Some | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='other lines are apart of text paragraphs. | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='First item. | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='N')",
    "MagicTodo(desc='High-priority item | @high ctime:HHMM dtime:HHMM', contexts=('high',), create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='A')",
    "MagicTodo(desc='Done item | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, done=True, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='B')",
    "MagicTodo(desc='Item is due:2022-01-31.', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='N')",
  ]
---
# name: test_group_from_path[no-filter-kwargs-foo]
  <class 'list'> [
    "MagicTodo(desc='I do lots of stuff in this file. Some lines are todo tasks, but some lines are headers. Some | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='other lines are apart of text paragraphs. | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    "MagicTodo(desc='First item. | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='N')",
    "MagicTodo(desc='High-priority item | @high ctime:HHMM dtime:HHMM', contexts=('high',), create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='A')",
    "MagicTodo(desc='High-priority item. | @high ctime:HHMM dtime:HHMM', contexts=('high',), create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='A')",
    "MagicTodo(desc='Done item | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, done=True, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='B')",
    "MagicTodo(desc='Item is due:2022-01-31.', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='N')",
    "MagicTodo(desc='This test item is awesome and has lots of | @stuff +test ctime:HHMM dtime:HHMM', contexts=('stuff',), create_date=1900-01-01, done_date=2022-02-07, done=True, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='Z', projects=('test',))",
    'MagicTodo(desc=\'Not a valid "strict" todo | ctime:HHMM dtime:HHMM\', create_date=1900-01-01, done_date=2022-02-07, metadata={\'ctime\': \'HHMM\', \'dtime\': \'HHMM\'})',
    'MagicTodo(desc=\'Done item with no priority is a valid "strict" todo | ctime:HHMM dtime:HHMM\', create_date=1900-01-01, done_date=2022-02-07, done=True, metadata={\'ctime\': \'HHMM\', \'dtime\': \'HHMM\'})',
    'MagicTodo(desc=\'A lowercase "o" can also be used when running in "strict" mode. | ctime:HHMM dtime:HHMM\', create_date=1900-01-01, done_date=2022-02-07, metadata={\'ctime\': \'HHMM\', \'dtime\': \'HHMM\'})',
    'MagicTodo(desc=\'But this "o" can only be used if a priority is NOT provided. | ctime:HHMM dtime:HHMM\', create_date=1900-01-01, done_date=2022-02-07, metadata={\'ctime\': \'HHMM\', \'dtime\': \'HHMM\'}, priority=\'N\')',
    'MagicTodo(desc="Don\'t + make + normal + pluses + projects. | ctime:HHMM dtime:HHMM", create_date=1900-01-01, done_date=2022-02-07, metadata={\'ctime\': \'HHMM\', \'dtime\': \'HHMM\'}, priority=\'N\')',
    "MagicTodo(desc='Ditto for keys: with no values. | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='N')",
    "MagicTodo(desc='We strip +punctuation. But not if the value:. is just @punctuation, ya:dig?', contexts=('punctuation',), create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='N', projects=('punctuation',))",
    "MagicTodo(desc='We allow descriptions to start with projects. | +project ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='N', projects=('project',))",
    "MagicTodo(desc='We allow descriptions to start with contexts. | @context ctime:HHMM dtime:HHMM', contexts=('context',), create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='N')",
    "MagicTodo(desc='1234 Todos can start with numbers:5678.', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'}, priority='N')",
    'MagicTodo(desc=\'Todos can easily log their "done times" with this trick. | ctime:HHMM dtime:HHMM\', create_date=1900-01-01, done_date=2022-02-07, done=True, metadata={\'ctime\': \'HHMM\', \'dtime\': \'HHMM\'})',
    "MagicTodo(desc='Double colons should be::ignored. | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
    'MagicTodo(desc="Sometimes we want to use apostrophe with project\'s or contexts\'. | @contexts +project ctime:HHMM dtime:HHMM", contexts=(\'contexts\',), create_date=1900-01-01, done_date=2022-02-07, metadata={\'ctime\': \'HHMM\', \'dtime\': \'HHMM\'}, projects=(\'project\',))',
    "MagicTodo(desc='We also ignore double ++pluses and double @@at signs. | ctime:HHMM dtime:HHMM', create_date=1900-01-01, done_date=2022-02-07, metadata={'ctime': 'HHMM', 'dtime': 'HHMM'})",
  ]
---
